Infrastructure additions

sethrj opened the pull request:
- Random number helper function and such
- Vector utilities
- Add cuda macros to SoftEqual

amandalund commented on src/random/distributions/GenerateCanonical.i.hh:
```diff
@@ -12,35 +12,34 @@ namespace celeritas
  */
-__device__ float GenerateCanonical<RngEngine, float>::operator()(RngEngine& rng)
+template<class RealType, class Generator>
+CELER_FUNCTION RealType sample_uniform(Generator& g)
```
```suggestion
CELER_FUNCTION RealType generate_canonical(Generator& g)
```

sethrj replied:
Uh oh, if this is still here how is everything building and passing? 😱 Good catch!

amandalund commented on src/random/distributions/GenerateCanonical.i.hh:
```diff
@@ -12,35 +12,34 @@ namespace celeritas
+CELER_FUNCTION RealType sample_uniform(Generator& g)
 {
-    return curand_uniform(rng.state_);
+    return GenerateCanonical<Generator, RealType>()();
```
```suggestion
    return GenerateCanonical<Generator, RealType>()(g);
```

amandalund commented on src/random/distributions/GenerateCanonical.i.hh:
```diff
@@ -12,35 +12,34 @@ namespace celeritas
-__device__ double
-GenerateCanonical<RngEngine, double>::operator()(RngEngine& rng)
+template<class Generator>
+CELER_FUNCTION real_type sample_uniform(Generator& g)
```
```suggestion
CELER_FUNCTION real_type generate_canonical(Generator& g)
```

amandalund commented on src/random/distributions/GenerateCanonical.i.hh:
```diff
@@ -12,35 +12,34 @@ namespace celeritas
+CELER_FUNCTION real_type sample_uniform(Generator& g)
 {
-    return curand_uniform_double(rng.state_);
+    return GenerateCanonical<Generator, real_type>()();
```
```suggestion
    return GenerateCanonical<Generator, real_type>()(g);
```

sethrj replied:
Clearly I must be missing a test if the code is building as-iss...

amandalund replied:
Haha yep, just those two helper functions are never used/tested.

amandalund commented on src/random/distributions/BinaryDistribution.hh:
```diff
@@ -0,0 +1,64 @@
+ * \note Since `true` is 1 and `false` is 0, the constructor argument is the
+ * _opposite_ order of std::discrete_distribution.
+ */
+class BinaryDistribution
```
Thoughts on naming this `BernoulliDistribution`?

sethrj replied:
Wow I'm a dum-dum, I didn't even know of that distribution's existence (or that a binary distribution is called a Bernoulli distribution). That's exactly what this should be. I'll update. Thanks!

sethrj commented:
OK @amandalund , everything's fixed now -- I changed all the distributions to use `generate_canonical` so that `GenerateCanonical` becomes an implementation detail (good for overriding with the various RNGs). I also updated the name of the distribution as you suggested. Finally, I realized I wasn't ever building with warnings on, so I changed my build flags to add warnings and fixed a couple of warnings that showed up.

sethrj commented:
Thanks @amandalund !
